### PR TITLE
feat: websocket support in webhooks

### PR DESCRIPTION
**Pull request by sudeshrshetty**
- websocket URLs are now supported in addition to existing HTTP webhook
URLs.
- Part of #1323

Signed-off-by: sudesh.shetty <sudesh.shetty@securekey.com>

### CODE REVIEW

File: pkg/controller/webhook/webhook.go [hunk 1]
```diff
@@ -89,6 +108,41 @@ func notify(destination string, message []byte) error {
 		destination, resp.Status)
 }
 
+func notifyWS(destination string, message []byte) error {
+	ctx, cancel := context.WithTimeout(context.Background(), notificationSendTimeout)
+	defer cancel()
+
+	c, rs, err := websocket.Dial(ctx, destination, nil)
```
Comment (troyronda) [inline pkg/controller/webhook/webhook.go]: So this is useful variation where the REST server dials a WebSocket...

But we really need a version where the REST server exposes a WebSocket and the client dials it. This way the client does not need an exposed endpoint.

File: pkg/controller/webhook/webhook.go [hunk 2]
```diff
@@ -89,6 +108,41 @@ func notify(destination string, message []byte) error {
...
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err = c.Close(websocket.StatusNormalClosure, "closing connection")
```
Comment (troyronda) [inline pkg/controller/webhook/webhook.go]: we should followup where we hold the connection open until an inactivity period.

File: pkg/controller/webhook/webhook.go (hunk 1 above)
Comment (troyronda) [reply pkg/controller/webhook/webhook.go]: > The implementation should work the same as it does in ACA-Py. The WebSocket support will allow for clients to register dynamically after the server has started. The full duplex functionality of WebSocket will not be used - it should be a one-way notification from server->client. The endpoint will be /ws.

File: pkg/controller/webhook/webhook.go [hunk 3]
```diff
@@ -30,20 +33,31 @@ type Notifier interface {
...
+type WebNotifier struct {
+	httpURLs []string
+	wsURLS   []string
 }
 
-// NewHTTPNotifier returns a new instance of an HTTPNotifier.
-func NewHTTPNotifier(webhookURLs []string) HTTPNotifier {
-	return HTTPNotifier{WebhookURLs: webhookURLs}
+// NewWebNotifier returns a new instance of an WebNotifier.
+func NewWebNotifier(webhookURLs []string) WebNotifier {
```
Comment (troyronda) [inline pkg/controller/webhook/webhook.go]: Should be returning a pointer.

File: pkg/controller/webhook/webhook.go [hunk 4]
```diff
@@ -89,6 +108,41 @@ func notify(destination string, message []byte) error {
 		destination, resp.Status)
 }
 
+func notifyWS(destination string, message []byte) error {
+	ctx, cancel := context.WithTimeout(context.Background(), notificationSendTimeout)
```
Comment (troyronda) [inline pkg/controller/webhook/webhook.go]: parent ctx should be an argument to Notify.

File: pkg/controller/webhook/webhook.go (hunk 1 above)
Comment (sudeshrshetty) [reply pkg/controller/webhook/webhook.go]: Will add that feature as part of separate PR

File: pkg/controller/webhook/webhook.go (hunk 2 above)
Comment (sudeshrshetty) [reply pkg/controller/webhook/webhook.go]: Added a follow up TODO

File: pkg/controller/webhook/webhook.go (hunk 3 above)
Comment (sudeshrshetty) [reply pkg/controller/webhook/webhook.go]: fixed

File: pkg/controller/webhook/webhook.go (hunk 4 above)
Comment (sudeshrshetty) [reply pkg/controller/webhook/webhook.go]: Added a follow up TODO and created an issue, applies to HTTP as well.

**Comment by sudeshrshetty** [comment]
pushed this as part of other PRs

**Pull request closed without merging**